src/hugr/views/petgraph.rs

```diff
@@ -89,91 +41,46 @@ impl<'g> pv::IntoNodeReferences for &'g Hugr {
     }
 }
 
-impl<'g> pv::IntoNeighbors for &'g Hugr {
-    type Neighbors = <Hugr as HugrView>::Neighbours<'g>;
-
-    fn neighbors(self, n: Self::NodeId) -> Self::Neighbors {
-        self.output_neighbours(n)
-    }
-}
-
-impl<'g> pv::IntoNeighborsDirected for &'g Hugr {
-    type NeighborsDirected = <Hugr as HugrView>::Neighbours<'g>;
-
-    fn neighbors_directed(
-        self,
-        n: Self::NodeId,
-        d: petgraph::Direction,
-    ) -> Self::NeighborsDirected {
-        self.neighbours(n, d.into())
-    }
-}
-
-impl pv::Visitable for Hugr {
-    type Map = std::collections::HashSet<Self::NodeId>;
-
-    fn visit_map(&self) -> Self::Map {
-        std::collections::HashSet::new()
-    }
-
-    fn reset_map(&self, map: &mut Self::Map) {
-        map.clear();
-    }
-}
-
-impl pv::GetAdjacencyMatrix for Hugr {
-    type AdjMatrix = std::collections::HashSet<(Self::NodeId, Self::NodeId)>;
+macro_rules! impl_petgraph_into_noderefs {
+    ($hugr:ident) => {
+        impl<'g, 'a, Root, Base> pv::IntoNodeReferences for &'g $hugr<'a, Root, Base>
+        where 
+        Root:NodeHandle, 'g: 'a, Base: HugrInternals + HugrView
+        {
+            type NodeRef = HugrNodeRef<'a>;
+            type NodeReferences =
+                MapWithCtx<<$hugr<'a, Root, Base> as HugrView>::Nodes<'a>, Self, HugrNodeRef<'a>>;
 
-    fn adjacency_matrix(&self) -> Self::AdjMatrix {
-        let mut matrix = std::collections::HashSet::new();
-        for node in self.nodes() {
-            for neighbour in self.output_neighbours(node) {
-                matrix.insert((node, neighbour));
+            fn node_references(self) -> Self::NodeReferences {
+                self.nodes()
+                    .with_context(self)
+                    .map_with_context(|n, &hugr| HugrNodeRef::from_node(n, hugr))
             }
         }
-        matrix
-    }
-
-    fn is_adjacent(&self, matrix: &Self::AdjMatrix, a: Self::NodeId, b: Self::NodeId) -> bool {
-        matrix.contains(&(a, b))
     }
 }
 
-impl PetgraphHugr for Hugr {}
-
 macro_rules! impl_region_petgraph_traits {
-    ($hugr:ident) => {
-        impl<'a, Root, Base> pv::GraphBase for $hugr<'a, Root, Base>
-        where
-            Root: NodeHandle,
-            Base: HugrInternals + HugrView,
+    ($hugr:ident $(< $($l:lifetime,)? $($v:ident: $b:tt $(+ $b2:tt)*),* >)? ) => {
+        impl $(<$($l,)? $($v: $b $(+ $b2)*),*>)? pv::GraphBase for $hugr$(<$($l,)? $($v),*>)?
```

Comment:
- Some generics may be unbounded. There should be an extra `$(: ...)?`

- `T<>` is valid syntax, so you could do
```rust
    ($hugr:ident) => {
        impl_region_petgraph_traits!{$hugr <>}
    };
    ($hugr:ident < $($l:lifetime,)? $($v:ident: $b:tt $(+ $b2:tt)*),* > ) => {
        ...
```
And then include the `'g` lifetime as needed between the always-present `<...>` for `IntoNodeReferences`.

Reply:
* Done. It was *not* a goal to support all possible type parameterizations (only ones we actually needed), but this isn't too hard.
* Nice! Yes, so `Hugr<>` is valid, and so  is `<'g,>`, that simplifies things a little. I needed a `#[rustfmt::skip]` to prevent `cargo fmt` changing `Hugr<>` to `Hugr`... and it throws away the comment in `Hugr< /* don't throw these away */>`!!

* However, `IntoNodeReferences` is a more complex problem - you need to insert `'g` if `$l` didn't match a lifetime, but use `$l` if it does. I felt there was enough difference between the instances that parameterizing the macro over those differences was not helpful rather than just giving both instances explicitly...